File: 2218-maximum-value-of-k-coins-from-piles/2218-maximum-value-of-k-coins-from-piles.py
class Solution:
    def maxValueOfCoins(self, piles: List[List[int]], k: int) -> int:
        
        @functools.lru_cache(None)
        def _search(rest_k: int, pile_pos: int) -> int:
            """Searches for best accumulated coin value starting from pile at position `pile_pos`.
            
            Args:
              - rest_k: The rest of coins we could pick.
              - pile_pos: The starting position of pile we are working on.
            """
            if rest_k == 0 or pile_pos == len(piles): return 0

            # Ignore the current pile and moving to next pile
            current_mv_found = _search(rest_k, pile_pos+1)

            # Start looking solution with current pile being involved
            current_pile = piles[pile_pos]
            accum_coin_value = 0
            for i in range(min(rest_k, len(current_pile))):
                # Accumulate the coins
                accum_coin_value += current_pile[i]

                # Moving to next pile with accumulated coin value in current pile
                current_mv_found = max(
                    current_mv_found, 
                    accum_coin_value + _search(rest_k-i-1, pile_pos+1))

            # Return the maximum accumulated coin value found from current pile.
            return current_mv_found

        # Start searching with initial `k` and first pile
        return _search(k, 0)

File: 2218-maximum-value-of-k-coins-from-piles/2218-maximum-value-of-k-coins-from-piles.cpp
class Solution {
public:
    int help(int i,vector<vector<int>>& piles,int k,vector<vector<int>>& dp){
        if(dp[i][k]>0)
            return dp[i][k];
        if(i==piles.size() or k==0)
            return 0;
        int res=help(i+1,piles,k,dp),curVal=0;
        for(int j=0;j<piles[i].size() and j<k;j++){
            curVal+=piles[i][j];
            res=max(res,help(i+1,piles,k-j-1,dp)+curVal);
        }
        dp[i][k]=res;
        return res;
    }
    int maxValueOfCoins(vector<vector<int>>& piles, int k) {
        int n=piles.size();
        vector<vector<int>> dp(n+1,vector<int>(k+1,0));
        return help(0,piles,k,dp);
    }
};